File: lib/rag/rerank.ts
// BGE-reranker via SiliconFlow's rerank endpoint.
// Docs: https://docs.siliconflow.cn/cn/api-reference/rerank/create-rerank
// Falls back to returning input unchanged if API key missing or request fails.

import type { RagHit } from './types'

export const RERANK_MODEL = process.env.RERANK_MODEL ?? 'BAAI/bge-reranker-v2-m3'
const BASE = process.env.SILICONFLOW_BASE_URL ?? 'https://api.siliconflow.cn/v1'

export async function rerank(query: string, hits: RagHit[], topK = 5): Promise<RagHit[]> {
  if (hits.length === 0) return []
  const key = process.env.SILICONFLOW_API_KEY
  if (!key) return hits.slice(0, topK)   // graceful fallback when key not configured

  try {
    const res = await fetch(`${BASE}/rerank`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${key}`,
      },
      body: JSON.stringify({
        model: RERANK_MODEL,
        query,
        documents: hits.map(h => h.text),
        top_n: Math.min(topK, hits.length),
        return_documents: false,
      }),
    })
    if (!res.ok) return hits.slice(0, topK)

    const json = await res.json() as { results: Array<{ index: number; relevance_score: number }> }
    const ranked = (json.results ?? [])
      .map(r => ({ ...hits[r.index], rerankScore: r.relevance_score }))
      .filter(Boolean)
    return ranked.slice(0, topK)
  } catch {
    return hits.slice(0, topK)
  }
}

File: server/stratsquad/rag/rerank.py
"""BGE reranker via SiliconFlow. Falls back to dense-only if key missing or call fails."""
from __future__ import annotations
import os
import httpx

from ..types import RagHit


RERANK_MODEL = os.getenv("RERANK_MODEL", "BAAI/bge-reranker-v2-m3")
BASE = os.getenv("SILICONFLOW_BASE_URL", "https://api.siliconflow.cn/v1")


async def rerank(query: str, hits: list[RagHit], top_k: int = 5) -> list[RagHit]:
    if not hits:
        return []
    key = os.getenv("SILICONFLOW_API_KEY")
    if not key:
        return hits[:top_k]
    try:
        async with httpx.AsyncClient(timeout=15) as client:
            res = await client.post(
                f"{BASE}/rerank",
                headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
                json={
                    "model": RERANK_MODEL,
                    "query": query,
                    "documents": [h.text for h in hits],
                    "top_n": min(top_k, len(hits)),
                    "return_documents": False,
                },
            )
            if res.status_code != 200:
                return hits[:top_k]
            results = res.json().get("results", [])
    except Exception:
        return hits[:top_k]

    out: list[RagHit] = []
    for r in results:
        idx = r.get("index")
        if idx is None or idx >= len(hits):
            continue
        h = hits[idx].model_copy(update={"rerank_score": r.get("relevance_score")})
        out.append(h)
    return out[:top_k]
